### PR TITLE
[feature] #2037: Introduce Pre-commit Triggers

### LINKED ISSUE #2037: Introduce pre-commit triggers
Pre-commit triggers are executed before every block commiit.
The implementation should be as simple as adding a new type of `TimeSchedule` for which `TimeFilter::count_matches()` returns `1` for every time event

### PR DESCRIPTION
**Pull request by Arjentix**
### Description of the Change

Initially I was planning to implement 3'rd test-case from #1890, but I forgot, that me and @appetrosyan decided to implement it with roles, not with triggers. But in the moment of realizing I've already finished this task, so here I am

- Some functions refactored to be const
- Added possibility to create pre-commit triggers as another kind of time-triggers
- Added integration test for pre-commit trigger

### Issue

Closes #2037 

### Benefits

Pre-commit triggers!

### Possible Drawbacks

None

### Usage Examples or Tests

See `client/tests/integration/triggers/time_trigger.rs`

### CODE REVIEW

**Comment by appetrosyan** [comment]
I'd like to comment that we could need pre-commit triggers for other purposes. For example for the atomic swap and offline transactions.

File: data_model/src/events/time.rs [hunk 1]
```diff
@@ -104,6 +109,29 @@ impl EventFilter {
...
+    PartialEq,
+    Eq,
+    Encode,
+    Decode,
+    Serialize,
+    Deserialize,
+    IntoSchema,
+    Hash,
+)]
+pub enum ExecutionTime {
```
Comment (appetrosyan) [inline data_model/src/events/time.rs:128]: This is an option-like enum. It should use the niche available in the other data type to encode the discriminant, but I would double check.

Comment (appetrosyan) [reply data_model/src/events/time.rs:128]: Yep. The sizes are identical even in Debug builds.

**Review by appetrosyan — APPROVED** [review]

**Review by s8sato — APPROVED** [review]
One question remains about its naming.
What is the reason we can say it is not post-commit but pre-commit?
It looks to me every-commit or per-commit

**Pull request merged**